### ISSUE TITLE
[BUG] lexical-core 0.6 pins cfg-if 0.1.9, which causes problems on newer stable.

### ISSUE THREAD

**Issue by jonhoo** [Alexhuszagh/rust-lexical#38]
## Description

`lexical-core 0.6` pins `cfg-if 0.1.9`, which causes downstream problems.
Users may be stuck on `lexical-core 0.6` for a while, since `nom 5` requires it, and moving to newer nom versions is a notoriously slow process.

## Prerequisites

- Rust version : `rustc 1.43.1 (8d69840ab 2020-05-04)`
- lexical-core version : `0.6.3`

## Test case

```toml
[dependencies]
lexical-core = "0.6" # or indirectly through nom = "5"
cfg-if = "0.1"
```

```rust
fn foo() {
    cfg_if::cfg_if! {
        if #[cfg(unix)] {
            fn bar() {}
            let tm = ();
        }
    }
}
```

```console
$ cargo +stable check
    Checking foo v0.1.0 (/home/jon/dev/tmp/foo)
error: expected an item keyword
 --> src/main.rs:5:13
  |
5 |             let tm = ();
  |             ^^^

error: aborting due to previous error

error: could not compile `foo`.
$ cargo +beta check
    Checking foo v0.1.0 (/home/jon/dev/tmp/foo)
error: expected an item keyword
  --> src/main.rs:5:13
   |
5  |             let tm = ();
   |             ^^^
   |
  ::: /home/jon/.cargo/registry/src/github.com-1ecc6299db9ec823/cfg-if-0.1.9/src/lib.rs:41:40
   |
41 |         if #[cfg($($meta:meta),*)] { $($it:item)* }
   |                                        -------- while parsing argument for this `item` macro fragment

error: aborting due to previous error

error: could not compile `foo`.
```

## Additional Context

So, the problem here is that cargo does a search of the dependency tree, sees that `lexical-core` requires exactly `cfg-if 0.1.9`, and so _any_ crate in the tree that depends on `cfg-if 0.1` then gets that version (since cargo only builds one major version of every crate).

As I understand it, the decision to pin `cfg-if 0.1.9` was made in https://github.com/Alexhuszagh/rust-lexical/commit/fefe81850e5678450ec0f001f562b182694caadf to support older Rust versions. Unfortunately that now means that _newer_ Rust versions are not supported. It seems more important to support new Rust versions than old ones, so I suggest that the pinning should be undone.

There is also a note in that PR saying:
> Update cfg-if to "0.1.10" when we support only Rustc >= 1.32.0.

Don't know if that applies now?

See also https://github.com/Geal/nom/pull/1115#issuecomment-632302676

**Comment by jonhoo** [comment]
This is particularly problematic because it causes compile issues that may not even be in the user's code. For example, any crate that _transitively_ depends on both `lexical-core 0.6` (e.g., through `nom 5`) and `time 0.2` will no longer compile on stable, with the error:
```
error: expected an item keyword
   --> /home/jon/.cargo/registry/src/github.com-1ecc6299db9ec823/time-0.2.8/src/utc_offset.rs:366:13
    |
366 |             let tm = timestamp_to_tm(datetime.timestamp())?;
    |             ^^^
    |
   ::: /home/jon/.cargo/registry/src/github.com-1ecc6299db9ec823/cfg-if-0.1.9/src/lib.rs:41:40
    |
41  |         if #[cfg($($meta:meta),*)] { $($it:item)* }
    |                                        -------- while parsing argument for this `item` macro fragment

error: aborting due to previous error
```

**Comment by jonhoo** [comment]
Note that newer versions of `time` all require `cfg-if ^0.1.10` for this reason (https://github.com/time-rs/time/pull/234). The `cfg-if 0.1.9` pin in this crate therefore forces `time` to be _downgraded_ to `0.2.8`, since that's the latest version that accepts `cfg-if 0.1.9`, but then that doesn't compile (https://github.com/time-rs/time/issues/233).

**Comment by jonhoo** [comment]
Gentle ping on this @Alexhuszagh — this is causing a bunch of pain for downstream crates.

**Comment by myrrlyn** [comment]
I am the new maintainer until the original Alex is able to return. It appears that we have been on `cfg-if` for several months now.

I am unfamiliar with `lexical`'s dependents. I have issued a `5.2.1` release of `lexical`, and `0.7.4` of `lexical-core`, but have not done anything else.

It appears that `nom` has updated to `0.7`. Is `0.6` still in use and requiring a backport?

**Comment by Alexhuszagh** [comment]
Does this still require a backport? I still commit to maintaining older versions, however, this was mostly for compatibility with older Rustc versions, which are now ~3 years old.

**Comment by jonhoo** [comment]
`nom 5` is still very much in use, in big part because upgrading a parser is a bunch of work. And looking at `cfg-if`'s [crates.io page](https://crates.io/crates/cfg-if), it seems about 1/3 of the uses of `cfg-if` are still 0.1. So yes, I'd still recommend a backport, as the original issue persists with just those two dependents:
```toml
[dependencies]
lexical-core = "*"
nom = "5"
cfg-if = "0.1"
```

**Comment by Alexhuszagh** [comment]
Will do, thank you.

**Comment by Alexhuszagh** [comment]
I'm looking forward to getting to this tomorrow, we've got a big release coming up for patches, and I've been preoccupied with other issues, but I should have time tomorrow. Thanks for your patience.

I will be backporting this to:
- v0.4
- v0.5
- v0.6

And that should cover any cases. Issues with cfg_if requiring new compiler versions aren't relevant anymore, since those are so ancient they're no longer relevant.

**Comment by Alexhuszagh** [comment]
[Tracking]
Currently backported the changes (along with other fixes) to:

Currently not yet implemented to:
- v0.4
- v0.5
- v0.6
- v0.8

**Comment by Alexhuszagh** [comment]
Closed as of [v0.4.8](https://github.com/Alexhuszagh/rust-lexical/releases/tag/v0.4.8), [v0.5.1](https://github.com/Alexhuszagh/rust-lexical/releases/tag/v0.5.1), [v0.6.8](https://github.com/Alexhuszagh/rust-lexical/releases/tag/v0.6.8), and [v0.7.6](https://github.com/Alexhuszagh/rust-lexical/releases/tag/v0.7.6).
